### PR TITLE
Add Enqueued Resources check

### LINKED ISSUE #25: Create Enqueued_Resources_Check
## Description

Checks that all plugin scripts and styles are enqueued using core functions and not directly output in HTML. This will use the existing WordPress.WP.EnqueuedResources sniff in the [WordPress Coding Standards](https://github.com/WordPress/WordPress-Coding-Standards/).

## Acceptance Criteria

- Class `Enqueued_Resources_Check` should be created and exists at `includes/Checker/Checks`
- Class `Enqueued_Resources_Check` should extend the `Abstract_PHP_CodeSniffer_Check` abstract class
- Class should implement the public `get_args()` and return an associative array of the required PHPCS arguments
    - `'sniffs' => 'WordPress.WP.EnqueuedResources'`
    - `'standard' => 'WordPress'`
- Add a sample `<link>` in the `tests/testdata/plugins/test-plugin-with-errors\test-plugin-with-errors.php` file to include any sample script. Make sure it will come under the `WordPress.WP.EnqueuedResources` error.

## Tests Coverage
- `get_args()` method returns the expected array of PHPCS arguments for the check
- `run()` method adds warnings to the `$check_result` when checking an invalid file - check reference [here](https://github.com/10up/plugin-check/blob/trunk/tests/Checker/Checks/I18n_Usage_Check_Tests.php#L20)
- `run()` method does not add any warnings to the `$check_result` when checking a valid file - check reference [here](https://github.com/10up/plugin-check/blob/trunk/tests/Checker/Checks/I18n_Usage_Check_Tests.php#L51)

### PR DESCRIPTION
**Pull request by ernilambar**
Fixes https://github.com/WordPress/plugin-check/issues/25

**Note:**

- I have added tests to cover my change.
- All new and existing tests pass.

### CODE REVIEW

**Review by swissspidy — APPROVED** [review]

File: tests/phpunit/testdata/plugins/test-plugin-enqueued-resources-with-errors/load.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+?>
+<link rel="stylesheet" href="http://someurl/somefile.css">
+<script src="http://someurl/somefile.js"></script>
+
+<?php
+
+?><link rel="stylesheet" href="https://someurl/somefile.css"><?php
+?><script src="https://someurl/somefile.js"></script><?php
```
Comment (mukeshpanchal27) [inline tests/phpunit/testdata/plugins/test-plugin-enqueued-resources-with-errors/load.php]: Single JS and CSS is enough no need to increase the error.

```suggestion
```

File: tests/phpunit/tests/Checker/Checks/Enqueued_Resources_Check_Tests.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		$errors = $check_result->get_errors();
+
+		$this->assertNotEmpty( $errors );
+		$this->assertArrayHasKey( 'load.php', $errors );
+		$this->assertEquals( 4, $check_result->get_error_count() );
+	}
+
+	public function test_run_without_errors() {
+		$check         = new Enqueued_Resources_Check();
+		$check_context = new Check_Context( UNIT_TESTS_PLUGIN_DIR . 'test-plugin-enqueued-resources-without-errors/load.php' );
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/Checker/Checks/Enqueued_Resources_Check_Tests.php:30]: The file lacks any code to verify the `Enqueued_Resources_Check`. To rectify this, it is advisable to enqueue both style and CSS within the file using the appropriate function. This action ensures that the test successfully passes.

**Review by mukeshpanchal27 — DISMISSED** [review]
Thank you for addressing the checks. I have provided some feedback for your consideration.

Regarding this matter, we are currently in the process of evaluating the acceptance criteria for this check. I don't believe it's fully prepared for engineering work yet. @felixarntz, could you please share your thoughts on this?

File: includes/Checker/Checks/Enqueued_Resources_Check.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	 * Gets the categories for the check.
+	 *
+	 * Every check must have at least one category.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array The categories for the check.
+	 */
+	public function get_categories() {
+		return array( Check_Categories::CATEGORY_PERFORMANCE );
```
Comment (felixarntz) [inline includes/Checker/Checks/Enqueued_Resources_Check.php]: Not sure about this. Whether or not you hard-code assets doesn't really do anything to performance. I think this is rather a general best practice or plugin requirement.

**Review by felixarntz — DISMISSED** [review]
@ernilambar @mukeshpanchal27 I think it might be okay to merge this as it's a simple check using WPCS and it's certainly a best practice. I would however prefer to check with the plugin review team whether they're okay with it.

I think in any case it could be merged as a warning, but right now it is an error and I would like to validate whether that's okay with the review team.

The other thing is that the category of this check should IMO be a different one.

File: includes/Checker/Checks/Enqueued_Resources_Check.php (hunk 3 above)
Comment (ernilambar) [reply includes/Checker/Checks/Enqueued_Resources_Check.php]: Which category is preferred here? I will update the PR when we get confirmation about warning/error from the review team.

**Comment by mukeshpanchal27** [comment]
> I think in any case it could be merged as a warning, but right now it is an error and I would like to validate whether that's okay with the review team.

The WPCS itself return error https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/WP/EnqueuedResourcesSniff.php#L65 so we should return error for this check. Please correct me if i miss anything. Thanks!

File: includes/Checker/Checks/Enqueued_Resources_Check.php (hunk 3 above)
Comment (mukeshpanchal27) [reply includes/Checker/Checks/Enqueued_Resources_Check.php]: I go with `General` category.

**Comment by ernilambar** [comment]
In the Acceptance Criteria in https://github.com/WordPress/plugin-check/issues/25 it is mentioned that warnings should be displayed.

**Comment by mukeshpanchal27** [comment]
It seems the AC are not currently up to date and haven't been approved by any team member. Perhaps this was mistakenly labeled as a notice when, in fact, the sniff itself returns an error, indicating that we should display an error message.

File: includes/Checker/Checks/Enqueued_Resources_Check.php (hunk 3 above)
Comment (ernilambar) [reply includes/Checker/Checks/Enqueued_Resources_Check.php]: Used `General` category.

**Comment by ernilambar** [comment]
PR has been rebased. `GENERAL` category is assigned for this check. Message is kept as `Error`.

**Comment by davidperezgar** [comment]
Hello, 
This check is part of Plugins Team, so the category should be Plugin repo. In our reviews, we actually enforce users to enqueue all assets, so it should be an error.

File: includes/Checker/Checks/Enqueued_Resources_Check.php (hunk 3 above)
Comment (davidperezgar) [reply includes/Checker/Checks/Enqueued_Resources_Check.php]: Hello, It has to be in repo category. It's a requirement for the plugins that we review.

Comment (ernilambar) [reply includes/Checker/Checks/Enqueued_Resources_Check.php]: @davidperezgar This has been already updated.

Comment (davidperezgar) [reply includes/Checker/Checks/Enqueued_Resources_Check.php]: Ok

**Comment by ernilambar** [comment]
PR has been updated and ready for review.

**Comment by davidperezgar** [comment]
Hello, I have compared Internal Scanner and this check and is not giving same results. 

For example is not detecting this code:
```
// Add the inline footer JS into an external JS file
function inline_footer_js() {
    // Create a nonce
    $nonce = wp_create_nonce('currency_converter_nonce');
    ?>
    <script type="text/javascript">
       var ajaxurl = '<?php echo esc_url(admin_url('admin-ajax.php')); ?>';
       var currencyConverterNonce = '<?php echo esc_js($nonce); ?>';
    </script>
    <?php
}
add_action('wp_footer', 'inline_footer_js');
```

What we are doing in our scanner is using regex. @frantorres could help as well on this. I give you the regex that we are using:
script_src: `/<script\s[^>]*?src\s*=\s*["|\']?([^>"\']*)["|\']?[^>]*?>/i`

script_type: `/<script\s[^>]*?type\s*=\s*["|\']?([^>"\']*(script)[^>"\']*)["|\']?[^>]*?>/i`

script_only: `/<script\s*>/i`

link: `/<link\s[^>]*?rel\s*=\s*["|\']?([^>"\']*(stylesheet)[^>"\']*)["|\']?[^>]*?>/i`

link-type: `/<link\s[^>]*?type\s*=\s*["|\']?([^>"\']*(text\\/css)[^>"\']*)["|\']?[^>]*?>/i`

style: `/<style(\s*|\s[^>]*)>/i`

They were tested and working correctly. I don't know the best approach, but WPCS seems not to be the only one.

**Comment by davidperezgar** [comment]
I've reviewed with Fran, and We could suggest to WPCS to have better detection of enqueues.

**Comment by swissspidy** [comment]
Yes, contributing to WPCS would be best here 👍 

Right now it only has 2 regexes for external scripts/styles, none for inline scripts etc.

Code in question: https://github.com/WordPress/WordPress-Coding-Standards/blob/d77b0204f49e4b4a2bbac791e3bbfbfe603edf06/WordPress/Sniffs/WP/EnqueuedResourcesSniff.php#L51-L81

**Review by davidperezgar — APPROVED** [review]
All seems ok. We're going to make an issue to WPCS in order to have better detection of enqueues.

**Pull request merged**